Postshader: Let shaders use the previous frame

unknownbrackets opened the pull request:
This is useful for i.e. simulating the slow update speed of the PSP's LCD screen, but could in theory be used for other effects.

Here's a test fragment shader I was using:

```glsl
#ifdef GL_ES
precision mediump float;
precision mediump int;
#endif

uniform sampler2D sampler0;
uniform sampler2D sampler2;
varying vec2 v_texcoord0;

void main() {
	vec3 update = texture2D(sampler0, v_texcoord0.xy).rgb;
	vec3 prev = texture2D(sampler2, v_texcoord0.xy).rgb;

	gl_FragColor.rgb = mix(prev, update, 0.1);
	gl_FragColor.a = 1.0;
}
```

Using `0.1` makes the effect very obvious, but of course that varies by rendered FPS.  For that reason I added a u_timeDelta:

Value | Meaning | Speeds up if unthrottled | Changes on load state
--- | --- | --- | ---
`u_timeDelta.x` | Fractional seconds, in real time, since last frame | No | No
`u_timeDelta.y` | Fractional seconds, in emulated time, since last frame (approx) | Yes | Yes
`u_timeDelta.z` | Number of vblanks (which occur 60 times every 1001ms, i.e. 59.94 times per second) since last frame | Yes | Yes
`u_timeDelta.w` | In practice, always 1.0 (delta number of flips) | N/A | N/A

With this, the above example can be amended with `uniform vec4 u_timeDelta;` at the top, and `u_timeDelta.y * 6` in place of 0.1 to have a more stable effect regardless of game FPS.  But you could also force 60 FPS output.

-[Unknown]

unknownbrackets committed:
Font: Prevent a crash on invalid data.

unknownbrackets committed:
Draw: Assert sampler bindings are in valid ranges.

unknownbrackets committed:
Postshader: Correct sampler binding translation.

unknownbrackets committed:
Postshader: Let shaders use the previous frame.

This is useful for i.e. simulating the slow update speed of the PSP's LCD
screen, but could in theory be used for other effects.

unknownbrackets committed:
Postshader: Add uniform for delta since last frame.

Useful mainly when using previous frame output.

hrydgard commented on Common/GPU/D3D11/thin3d_d3d11.cpp:
```diff
@@ -1329,6 +1332,7 @@ Framebuffer *D3D11DrawContext::CreateFramebuffer(const FramebufferDesc &desc) {
 void D3D11DrawContext::BindTextures(int start, int count, Texture **textures) {
 	// Collect the resource views from the textures.
 	ID3D11ShaderResourceView *views[8];
```
might as well change this array length to MAX_BOUND_TEXTURES since you're here poking around

unknownbrackets replied:
Yep, agreed.  Went ahead and cleaned up some other GL checks in Draw too along the same lines.

-[Unknown]

hrydgard commented on Common/GPU/D3D11/thin3d_d3d11.cpp:
```diff
@@ -1338,6 +1342,7 @@ void D3D11DrawContext::BindTextures(int start, int count, Texture **textures) {
 
 void D3D11DrawContext::BindSamplerStates(int start, int count, SamplerState **states) {
 	ID3D11SamplerState *samplers[8];
```
same

hrydgard reviewed (commented):
Nice refactoring too! Just nitpicking.

unknownbrackets committed:
Draw: Use constants for texture slots.

GL actually had a check for 16 but then an array for 8.  This should make
it easier to figure out if we ever hit those limits.

LunaMoo commented:
Not sure if I'm doing something wrong, but even with a test shader posted here past frame is always black, so the test shader results in a 90% mix of black and 10% of current frame, so just a very dark image from current frame.
 Does this have some special requirements? I have the same results on all backends.

iota97 commented:
> Does this have some special requirements?

Have you set `UsePreviousFrame=true` in the ini?

LunaMoo commented:
Aw, I completely missed that, it works fine now;3, thanks.
